src/utils/config.c

```diff
@@ -357,3 +362,113 @@ bool conf_is_list(config_setting_t *conf, const char *file, size_t line,
     }
     return true;
 }
+
+/*
+ * config_group_apply
+ *      Apply a function to a group. Optional `arg` can be used to pass user
+ *      data into the specified function.
+ */
+void config_group_apply(const config_setting_t *options, group_func func, void *arg)
+{
+    int noptions;
+
+    if (!options)
+        return;
+
+    noptions = config_setting_length(options);
+
+    for (int i = 0; i < noptions; ++i)
+    {
+        config_setting_t *o;
+        const char *key;
+        const char *value;
+
+        o = config_setting_get_elem(options, i);
+        key = config_setting_name(o);
+        value = config_setting_get_string(o);
+
+        func(key, value, arg);
+    }
+}
+
+/*
+ * config_create_path
+ *      Creates path nodes except the leaf one. The path must be specified
+ *      in the form "one/two/three", e.i. path parts are separated with the '/'
+ *      symbol
+ */
+config_setting_t *config_create_path(config_setting_t *parent,
+                                     const char *path,
+                                     int type)
+{
+    char   *sep;
+    config_setting_t *cur = parent;
+    char    buf[512];
+    char   *prefix = buf;
+
+    assert(strlen(path) < 512);
+    strncpy(buf, path, sizeof(buf));
+    buf[sizeof(buf) - 1] = '\0';
+
+    while ((sep = strchr(prefix, PATH_SEPARATOR)) != NULL)
+    {
+        config_setting_t *found;
+        *sep = '\0';
+
+        /* lookup name in the current group */
+        found = config_setting_lookup(cur, prefix);
+
+        if (!found)
+            cur = config_setting_add(cur, prefix, CONFIG_TYPE_GROUP);
```

Comment:
Technically we'd need to check for null here, config_setting_add would probably segfault later. (It is exceedingly unlikely nothing else fails on this unhappy path though).